examples/Nat/NATBase.cs

```diff
@@ -0,0 +1,287 @@
+      packet.SetSource(connection.NatNode);
+
+      // Update checksums
+      packet.UpdateChecksums();
+
+      // Forward on the mapped network port
+      return new ForwardingDecision.SinglePortForward(connection.OutsideNode.InterfaceNumber);
+    }
+
+    /// <summary>
+    /// Remove connections that have timed out or are closed.
+    /// </summary>
+    public void GarbageCollectConnections()
+    {
+      bool removedAny = false;
+      DateTime now = DateTime.Now;
+      foreach (var pair in NAT_MapToInside)
+      {
+        NatConnection<TPacket,TNode> connection = pair.Value;
+        bool removeEntry = false;
+        if (now - connection.LastUsed > InactivityTimeout)
+        {
+          removeEntry = true;
+#if DEBUG
+          Console.WriteLine("Removing inactive connection (LastUsed {0}, Diff {1}, InactivityTimeout {2})",
+            connection.LastUsed.ToShortTimeString(),
+            (now - connection.LastUsed).ToString(),
+            InactivityTimeout);
+#endif
+        }
+        else if (connection.State.CanBeClosed)
+        {
+          removeEntry = true;
+#if DEBUG
+          Console.WriteLine("Removing closed connection (LastUsed {0}, Diff {1}, InactivityTimeout {2})",
+            connection.LastUsed.ToShortTimeString(),
+            (now - connection.LastUsed).ToString(),
+            InactivityTimeout);
+#endif
+        }
+
+        if (removeEntry)
+        {
+          // Remove this connection from both lookups
+          NAT_MapToInside.Remove(pair);
+          NAT_MapToOutside.Remove(new ConnectionKey(connection.InsideNode, connection.OutsideNode));
+#if DEBUG
+          removedAny = true;
+#endif
+        }
+      }
+
+#if DEBUG
+      if (removedAny)
+        PrintMappings();
+#endif
+
+      // FIXME - currently O(n); improve performance for sparse traffic.
+      // Could use a regularly invoked GC, and keep the connections in one of two pools:
+      //   - For connections with lower traffic, keep a double-LL of MRU, moving the connection to the head when used,
+      //     and start at the LRU end when iterating. Needs to be concurrent.
+      //     Cost is per packet, and maintains an ordering, meaning we only need to check those that have expired plus one extra.
+      //   - For connections with higher traffic, it would be less expensive to leave unsorted and iterate in O(n).
+      // However, this would add complexity in terms of tracking which pool each connection should be in.
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="TNode"/> that can act as the masquerading address for a new connection. E.g. an unused TCP socket.
+    /// </summary>
+    /// <param name="ipAddress">The IP address to use.</param>
+    /// <param name="interfaceNumber">The interface number to use.</param>
+    /// <param name="macAddress">The MAC address to use.</param>
+    protected abstract TNode CreateMasqueradeNode(IPAddress ipAddress, int interfaceNumber, PhysicalAddress macAddress);
```

Comment:
I noticed that `TNode` is being used both as a name of an abstract class, and as a generic parameter (e.g., in `NatConnection`). It would be preferable to avoid this shadowing.

Also it's not immediately clear what the purpose of `TNode` is -- could you elaborate a bit in the description please?